src/dataflow/src/render/delta_join.rs

```diff
@@ -341,146 +382,224 @@ where
-                {
-                    // Remove expression and equate with the first bound expression.
-                    should_equate.push(equivalence.remove(cursor));
-                } else {
-                    cursor += 1;
+impl MyClosure {
+    /// Applies per-row filtering and logic.
+    #[inline(always)]
+    pub fn apply<'a>(
+        &'a self,
+        datums: &mut Vec<Datum<'a>>,
+        temp_storage: &'a RowArena,
+        row_packer: &mut RowPacker,
+    ) -> Result<Option<Row>, expr::EvalError> {
+        for exprs in self.ready_equivalences.iter() {
+            // Each list of expressions should be equal to the same value.
+            let val = exprs[0].eval(&datums[..], &temp_storage)?;
+            for expr in exprs[1..].iter() {
+                if expr.eval(&datums, &temp_storage)? != val {
+                    return Ok(None);
                 }
             }
-            if !should_equate.is_empty() {
-                should_equate.push(equivalence[pos].clone());
-                ready_equivalences.push(should_equate);
-            }
         }
+        self.before.evaluate(datums, &temp_storage, row_packer)
     }
-    equivalences.retain(|e| e.len() > 1);
 
-    // Rewrite column references to their locations under `source_columns`.
-    for expr in ready_to_go.iter_mut() {
-        expr.visit_mut(&mut |e| {
-            if let ScalarExpr::Column(c) = e {
-                *c = source_columns
-                    .iter()
-                    .position(|x| x == c)
-                    .expect("Column not found in source_columns");
-            }
-        })
-    }
-    for exprs in ready_equivalences.iter_mut() {
-        for expr in exprs.iter_mut() {
-            expr.visit_mut(&mut |e| {
-                if let ScalarExpr::Column(c) = e {
-                    *c = source_columns
-                        .iter()
-                        .position(|x| x == c)
-                        .expect("Column not found in source_columns");
+    /// Construct an instance of the closue from available columns.
+    ///
+    /// This method updates the available columns, equivalences, and
+    /// the `MapFilterProject` instance. The columns are updated to
+    /// include reference to any columns added by the application of
+    /// this logic, which might result from partial application of
+    /// the `MapFilterProject` instance.
+    ///
+    /// If all columns are available for `mfp`, this method works
+    /// extra hard to ensure that the closure contains all the work,
+    /// and `mfp` is left as an identity transform (which can then
+    /// be ignored).
+    pub fn build(
+        columns: &mut HashMap<usize, usize>,
+        equivalences: &mut Vec<Vec<ScalarExpr>>,
+        mfp: &mut MapFilterProject,
+    ) -> Self {
+        // First, determine which columns should be compare due to `equivalences`.
+        let mut ready_equivalences = Vec::new();
+        for equivalence in equivalences.iter_mut() {
```

Comment:
It took me a while to grok that the body of this for loop + the `retain` were essentially computing a `drain_filter`

I think this diff is simpler?

```
diff --git a/src/dataflow/src/render/delta_join.rs b/src/dataflow/src/render/delta_join.rs
index d2322e0a3..a47cbacf3 100644
--- a/src/dataflow/src/render/delta_join.rs
+++ b/src/dataflow/src/render/delta_join.rs
@@ -497,32 +497,24 @@ impl MyClosure {
         // First, determine which columns should be compare due to `equivalences`.
         let mut ready_equivalences = Vec::new();
         for equivalence in equivalences.iter_mut() {
-            if let Some(pos) = equivalence
-                .iter()
-                .position(|e| e.support().into_iter().all(|c| columns.contains_key(&c)))
-            {
-                let mut should_equate = Vec::new();
-                let mut cursor = pos + 1;
-                while cursor < equivalence.len() {
-                    if equivalence[cursor]
-                        .support()
-                        .into_iter()
-                        .all(|c| columns.contains_key(&c))
-                    {
-                        // Remove expression and equate with the first bound expression.
-                        should_equate.push(equivalence.remove(cursor));
-                    } else {
-                        cursor += 1;
-                    }
-                }
-                if !should_equate.is_empty() {
-                    should_equate.push(equivalence[pos].clone());
-                    ready_equivalences.push(should_equate);
+            let mut cursor = 0;
+            let mut should_equate = Vec::new();
+            while cursor != equivalence.len() {
+                if equivalence[cursor]
+                    .support()
+                    .into_iter()
+                    .all(|c| columns.contains_key(&c))
+                {
+                    // Remove expression and equate with the first bound expression.
+                    should_equate.push(equivalence.remove(cursor));
+                } else {
+                    cursor += 1;
                 }
             }
+            if !should_equate.is_empty() {
+                ready_equivalences.push(should_equate);
+            }
         }
-        equivalences.retain(|e| e.len() > 1);
-
         // Update ready_equivalences to reference correct column locations.
         for exprs in ready_equivalences.iter_mut() {
             for expr in exprs.iter_mut() {
```

i checked that it compiles but i didnt check for correctness

Reply:
The initial version is roughly copy/paste from `build_filter` which was the code that did this same thing the first time around (it gets removed in one of the commits, too). I *think* your version removes all equated terms, rather than all but the first (we need to leave the first around, as it may be equated to more things in the future). Happy to try and simplify, but also a little hesitant! :D

Reply:
oh I see we might need to keep the first around if there was another one that doesn't match the predicate (if not all columns are present), in which case the retain won't kick in. that makes sense

no worries on the hesitation, part of trying to do the rewrite was just to get a better handle on the code :)